### PR TITLE
Adding support for multiple contracts

### PR DESCRIPTION
**Pull request by matthewjacobson**
Adding support for multiple contracts. Had to change the .env file a bit - now passing all contract info as a single json string. Not a very elegant solution but I couldn't think of anything better. It would look like this for a plottables deployment:

```
{
    "0xa319c382a702682129fcbf55d514e61a16f97f9c": {
        "MINT_CONTRACT_ADDRESS": "0x463b8ced7d22a55aa4a5d69ef6a54a08aa0feb93",
        "MEDIA_URL": "https://plottables-mainnet.s3.amazonaws.com",
        "TOKEN_URL": "https://token.artblocks.io",
        "GENERATOR_URL": "https://generator.artblocks.io"
    },
    "0x18de6097ce5b5b2724c9cae6ac519917f3f178c0": {
        "MINT_CONTRACT_ADDRESS": "0xe6e728361b7c824cba64cc1e5323efb7a5bb65da",
        "MEDIA_URL": "https://plottables-flex-mainnet.s3.amazonaws.com",
        "TOKEN_URL": "https://token.artblocks.io",
        "GENERATOR_URL": "https://generator.artblocks.io"
    }
}
```

The TOKEN_URL and GENERATOR_URL per project is probably redundant but figured it wouldn't hurt to throw them in there.

Future Additions:
- Minting support for older contract versions - would probably require a new field in the REACT_APP_CONTRACT_INFO .env variable indicating the version when deploying for contracts spanning multiple versions.
- A User page listing all tokens owned by a wallet

Issues:
- On the /projects page the pagination is kind of broken - but only when trying to go to a page number lower than the one currently selected. Need to find a better way to get the total number of projects - currently using max project id which breaks when using multiple contracts.

### CODE REVIEW

**Comment by matthewjacobson** [comment]
Added migration to latest versions of rainbowkit and wagmi to avoid the issue mentioned at https://github.com/rainbow-me/rainbowkit/issues/837

**Comment by matthewjacobson** [comment]
The issue I mentioned earlier on the /projects page isn't actually related to pagination (as far as I can tell). Seems to be a bug in mui masonry as mentioned at https://stackoverflow.com/questions/75774800/how-to-stop-resizeobserver-loop-limit-exceeded-error-from-appearing-in-react-a

However I do still think using the max project id will be a problem

File: sample.env [hunk 1]
```diff
@@ -1,8 +1,4 @@
 REACT_APP_EXPECTED_CHAIN_ID=1
 REACT_APP_INFURA_KEY=
 REACT_APP_GRAPHQL_URL=https://api.thegraph.com/subgraphs/name/artblocks/art-blocks
-REACT_APP_CORE_CONTRACT_ADDRESS=0x99a9b7c1116f9ceeb1652de04d5969cce509b069
-REACT_APP_MINT_CONTRACT_ADDRESS=0x706d6C6ef700a3c1C3a727f0c46492492E0A72b5
-REACT_APP_MEDIA_URL=https://media.artblocks.io
-REACT_APP_TOKEN_URL=https://token.artblocks.io
-REACT_APP_GENERATOR_URL=https://generator.artblocks.io
+REACT_APP_CONTRACT_INFO={"0x99a9b7c1116f9ceeb1652de04d5969cce509b069":{"MINT_CONTRACT_ADDRESS":"0x706d6C6ef700a3c1C3a727f0c46492492E0A72b5","MEDIA_URL":"https://media.artblocks.io","TOKEN_URL":"https://token.artblocks.io","GENERATOR_URL":"https://generator.artblocks.io"}}
```
Comment (grantoesterling) [inline sample.env]: Could this be defined in a .json config file instead of the .env file ([a la artbot](https://github.com/ArtBlocks/artbot/blob/main/src/ProjectConfig/channels.json))? seems like that would be much more readable/sustainable, plus i don't think this info is confidential or anything

File: src/components/Header.tsx [hunk 2]
```diff
@@ -19,11 +19,6 @@ const items = [
     label: "Projects",
     url: "/projects",
     enabled: true
-  },
-  {
-    label: "Mint",
```
Comment (grantoesterling) [inline src/components/Header.tsx:24]: Do you think this should be kept in as an example of a disabled / potential tab @jakerockland ?

File: src/components/ProjectPreview.tsx [hunk 3]
```diff
@@ -24,7 +24,7 @@ const ProjectPreview = ({project, width=280, showDescription=false}: Props) => {
   return (
     <Box>
       <Box>
-        <Link href={`/project/${project.projectId}`} underline="hover">
+        <Link href={`/project/${project.id.split('-')[0]}/${project.projectId}`} underline="hover">
```
Comment (grantoesterling) [inline src/components/ProjectPreview.tsx]: I think it makes sense to add the `contract` field to the `GetProject` query in `useProject.tsx` in order to do `project.contract.address` instead of the splitting here

File: src/components/TokenImage.tsx [hunk 4]
```diff
@@ -7,9 +7,10 @@ interface Props {
 }
 
 const TokenImage = ({tokenId, width, height}: Props) => {
+  const mediaUrl = CONTRACT_INFO[tokenId.split('-')[0]].MEDIA_URL
   return (
     <img
-      src={`${MEDIA_URL}/${tokenId}.png`}
+      src={`${mediaUrl}/${tokenId.split('-')[1]}.png`}
```
Comment (grantoesterling) [inline src/components/TokenImage.tsx]: why does this split need to happen? is the tokenId that is passed in different now?

File: src/components/TokenLive.tsx [hunk 5]
```diff
@@ -20,7 +20,8 @@ const TokenLive = ({tokenId, width, height}: Props) => {
   const [pollingTime, setPollingTime] = useState(0)
   const [pollingDelay, setPollingDelay] = useState(0)
   const [pollingAttempts, setPollingAttempts] = useState(0)
-  const endpoint = `${GENERATOR_URL}/${CORE_CONTRACT_ADDRESS?.toLowerCase()}/${tokenId}`
+  const generatorUrl = CONTRACT_INFO[tokenId.split('-')[0]].GENERATOR_URL
+  const endpoint = `${generatorUrl}/${tokenId.split('-')[0].toLowerCase()}/${tokenId.split('-')[1]}`
```
Comment (grantoesterling) [inline src/components/TokenLive.tsx]: ditto the above, we can probs add the contract field to the GetToken query instead of splitting all these places, for readability reasons

File: src/components/TokenView.tsx [hunk 6]
```diff
@@ -36,10 +35,10 @@ const TokenView = ({
           )
         }
       </Card>
-      { invocation !== undefined && 
+      { invocation !== undefined &&
         (
         <Box sx={{width: "100%", display: "flex", justifyContent: "flex-end"}}>
-          <Link href={`/token/${CORE_CONTRACT_ADDRESS}-${tokenId}`} sx={{fontSize: "14px", textDecoration: "none"}}>
+          <Link href={`/token/${tokenId.split('-')[0]}/${tokenId.split('-')[1]}`} sx={{fontSize: "14px", textDecoration: "none"}}>
```
Comment (grantoesterling) [inline src/components/TokenView.tsx]: ditto above on the splitting

File: src/components/Tokens.tsx [hunk 7]
```diff
@@ -63,14 +63,14 @@ const Tokens = ({
...
 
   return (
-    data.tokens.length > 0 ? 
+    data.tokens.length > 0 ?
       <Grid spacing={2} container>
         {
           data.tokens.map(((token:Token) => (
             <Grid key={token.tokenId} item md={4} sm={6} xs={6}>
-              <Link href={`/token/${token.id}`}>
+              <Link href={`/token/${token.id.split('-')[0]}/${token.id.split('-')[1]}`}>
```
Comment (grantoesterling) [inline src/components/Tokens.tsx]: splitting

File: src/hooks/useGeneratorPreview.tsx [hunk 8]
```diff
@@ -28,7 +28,8 @@ const useGeneratorPreview = (project: Project) => {
     setLoading(true)
     try {
       const token = generateToken(Number(project.projectId))
-      const { data } = await axios.get(`${GENERATOR_URL}/${project.id}/${token.tokenId}/${token.hash}`)
+      const generatorUrl = CONTRACT_INFO[project.id.split('-')[0]].GENERATOR_URL
```
Comment (grantoesterling) [inline src/hooks/useGeneratorPreview.tsx]: splitting

**Review by grantoesterling — DISMISSED** [review]
looks good! just a few comments / questions :)

File: sample.env (hunk 1 above)
Comment (jakerockland) [reply sample.env]: ++

File: src/components/Header.tsx (hunk 2 above)
Comment (jakerockland) [reply src/components/Header.tsx:24]: That seems reasonable to me. @Jbern16 what are your thoughts here?

File: sample.env (hunk 1 above)
Comment (Jbern16) [reply sample.env]: yes please!

File: src/components/Header.tsx (hunk 2 above)
Comment (Jbern16) [reply src/components/Header.tsx:24]: Seems fine, generally a fan of more examples the better

File: src/components/ProjectDetails.tsx [hunk 9]
```diff
@@ -75,7 +76,7 @@ const ProjectDetails = ({ id }: Props) => {
           token && (
             <Grid item md={8}>
               <TokenView
-                tokenId={token.tokenId}
+                tokenId={token.id}
```
Comment (Jbern16) [inline src/components/ProjectDetails.tsx]: don't think we need to change this and instead use the token.contract_address instead where we need the contract addy

**Review by Jbern16 — APPROVED** [review]
Generally LGTM but ditto grants comments!

**Comment by matthewjacobson** [comment]
Thanks so much for the feedback! Going to start addressing some of your comments.

File: sample.env (hunk 1 above)
Comment (matthewjacobson) [reply sample.env]: Agreed this is a great idea. Addressed with [49e7b63](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/49e7b639877e81958340e0d58e1095c25ed2aaf2)

File: src/components/Header.tsx (hunk 2 above)
Comment (matthewjacobson) [reply src/components/Header.tsx:24]: address with [59d2ffa](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/59d2ffa1d07b84eb86856625d2570d9849365f52)

File: src/hooks/useGeneratorPreview.tsx (hunk 8 above)
Comment (matthewjacobson) [reply src/hooks/useGeneratorPreview.tsx]: addressed with [49e7b63](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/49e7b639877e81958340e0d58e1095c25ed2aaf2)

File: src/components/ProjectPreview.tsx (hunk 3 above)
Comment (matthewjacobson) [reply src/components/ProjectPreview.tsx]: addressed with [cc69df2](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/cc69df2d9aca3b7a9ce1e7f4dfd30fbc722eca3f)

File: src/components/ProjectDetails.tsx (hunk 9 above)
Comment (matthewjacobson) [reply src/components/ProjectDetails.tsx]: addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

File: src/components/Tokens.tsx (hunk 7 above)
Comment (matthewjacobson) [reply src/components/Tokens.tsx]: addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

File: src/components/TokenView.tsx (hunk 6 above)
Comment (matthewjacobson) [reply src/components/TokenView.tsx]: addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

File: src/components/TokenLive.tsx (hunk 5 above)
Comment (matthewjacobson) [reply src/components/TokenLive.tsx]: addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

File: src/components/TokenImage.tsx (hunk 4 above)
Comment (matthewjacobson) [reply src/components/TokenImage.tsx]: addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

**Comment by jakerockland** [comment]
Re-requesting review from your end @grantoesterling

File: src/components/TokenDetails.tsx [hunk 10]
```diff
@@ -22,14 +22,18 @@ import useToken from "hooks/useToken"
...
 
-const TokenDetails = ({ id }: Props) => {
+const TokenDetails = ({ contractAddress, id }: Props) => {
   const theme = useTheme()
   const windowSize = useWindowSize()
-  const { loading, error, data } = useToken(id)
+  const { loading, error, data } = useToken(`${contractAddress}-${id}`)
   const token = data?.token
+  const contractConfig = CONTRACT_INFO.filter(
+      x => x.CORE_CONTRACT_ADDRESS.toLowerCase() == contractAddress.toLowerCase()
```
Comment (grantoesterling) [inline src/components/TokenDetails.tsx]: nit: I think this should be `===`

File: src/components/TokenDetails.tsx [hunk 11]
```diff
@@ -86,7 +91,8 @@ const TokenDetails = ({ id }: Props) => {
                   padding: [0, 0, "default"]
                 }}
                 onClick={() => {
-                  window.open(`${GENERATOR_URL}/${CORE_CONTRACT_ADDRESS?.toLowerCase()}/${token.tokenId}`)
+                  const generatorUrl = contractConfig[0].GENERATOR_URL
```
Comment (grantoesterling) [inline src/components/TokenDetails.tsx]: is this supposed to be hardcoded as 0?

Comment (grantoesterling) [reply src/components/TokenDetails.tsx]: I suppose it doesn't matter since the links are the same, but for robustness' sake should probably make this programatic depending on the token (in a world where someone might have 2 diff generator urls). A util function of some sort maybe where you pass in contractAddress and get generatorUrl/mediaUrl/etc?

Alternatively if generator link will ALWAYS be the same for everyone, maybe it is just its own config var instead of being in that contractConfig struct

File: src/components/TokenDetails.tsx [hunk 12]
```diff
@@ -103,7 +109,8 @@ const TokenDetails = ({ id }: Props) => {
                   padding: [0, 0, "default"]
                 }}
                 onClick={() => {
-                  window.open(`${MEDIA_URL}/${token.tokenId}.png`)
+                  const mediaUrl = contractConfig[0].MEDIA_URL
```
Comment (grantoesterling) [inline src/components/TokenDetails.tsx]: ditto the above

File: src/components/TokenImage.tsx [hunk 13]
```diff
@@ -1,15 +1,20 @@
...
-const TokenImage = ({tokenId, width, height}: Props) => {
+const TokenImage = ({contractAddress, tokenId, width, height}: Props) => {
+  const contractConfig = CONTRACT_INFO.filter(
+      x => x.CORE_CONTRACT_ADDRESS.toLowerCase() == contractAddress.toLowerCase()
+  )
+
   return (
     <img
-      src={`${MEDIA_URL}/${tokenId}.png`}
+      src={`${contractConfig[0].MEDIA_URL}/${tokenId}.png`}
```
Comment (grantoesterling) [inline src/components/TokenImage.tsx]: ditto hardcoded index comment

File: src/components/TokenDetails.tsx (hunk 12 above)
Comment (grantoesterling) [reply src/components/TokenDetails.tsx]: Actually this one definitely shouldn't be hardcoded right? since media urls will be different?

**Review by grantoesterling — APPROVED** [review]
LGTM other than the question about hardcoded indicies!

File: src/components/TokenDetails.tsx (hunk 12 above)
Comment (Jbern16) [reply src/components/TokenDetails.tsx]: media_url is this context I believe might be referencing our media proxy? However, you should be able to get the media_url from the tokens_metadata. `token.media_url`

**Comment by jakerockland** [comment]
Thank you very much @grantoesterling @Jbern16 for the thoughtful reviews here.

Thank you @matthewjacobson for sending along this PR !! 💜 🙏

File: src/components/TokenDetails.tsx (hunk 10 above)
Comment (matthewjacobson) [reply src/components/TokenDetails.tsx]: Addressed with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c) in utils/contractInfoHelper.ts

File: src/components/TokenDetails.tsx (hunk 11 above)
Comment (matthewjacobson) [reply src/components/TokenDetails.tsx]: The 0 hardcoding was intentional (although probably bad practice) - I was using CONTRACT_INFO.filter() to get the proper contract config, and this returned an array for which there should have been only a single element. I have addressed this with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c). Now using CONTRACT_INFO.find() to avoid the hardcoding.

File: src/components/TokenDetails.tsx (hunk 12 above)
Comment (matthewjacobson) [reply src/components/TokenDetails.tsx]: Address with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c)

@grantoesterling yes the media url will be different for each contract - at least that has been my experience with plottables.

@Jbern16 I'm not seeing the media_url in the tokens_metadata but maybe I am missing something. Is this coming from the subgraph somewhere?

File: src/components/TokenImage.tsx (hunk 13 above)
Comment (matthewjacobson) [reply src/components/TokenImage.tsx]: Address with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c)

**Comment by matthewjacobson** [comment]
> Thank you very much @grantoesterling @Jbern16 for the thoughtful reviews here.
> 
> Thank you @matthewjacobson for sending along this PR !! 💜 🙏

Thank you! And thanks to @grantoesterling and @Jbern16 for the reviews!

Two other things we could maybe address before pulling:
1) Do any of you have thoughts on the best way to add support for the older V2 contracts? Would it make sense to add a new MintingInterfaceV2 component? And then do something like this in ProjectDetails.tsx:
```
{
  {
    "V3": <MintingInterfaceV3 ... />,
    "V2": <MintingInterfaceV2 ... />
  }
}[contractConfig?.VERSION]
```
2) Are you all experiencing the same issue I am on the /projects page? Resizing the page (or going from page 2 back to page 1 throws a "ResizeObserver loop limit exceeded" error. As I mentioned earlier this seems to be a [bug in mui masonry](https://stackoverflow.com/questions/75774800/how-to-stop-resizeobserver-loop-limit-exceeded-error-from-appearing-in-react-a). Should I update this to use mui grid?

File: src/components/TokenDetails.tsx (hunk 12 above)
Comment (Jbern16) [reply src/components/TokenDetails.tsx]: ah - my apologies, this is from our off chain api. We should be opening that up to the public soon, but perhaps until then we can hardcode the endpoints as it might streamline this logic - 

```
https://media-proxy.artblocks.io/0x54a6356244059d5a50b97200a928f19a3682b669/10
https://media-proxy-staging.artblocks.io/0xda62f67be7194775a75be91cbf9feedcc5776d4b/97000751
https://media-proxy-development.artblocks.io/0x5702797ff45fcb0a70eb6ae1e4563299dcfa9dd6/1

{contract_address}/{token_id}
```

**Comment by Jbern16** [comment]
@matthewjacobson 

> > Thank you very much @grantoesterling @Jbern16 for the thoughtful reviews here.
> > Thank you @matthewjacobson for sending along this PR !! 💜 🙏
> 
> Thank you! And thanks to @grantoesterling and @Jbern16 for the reviews!
> 
> Two other things we could maybe address before pulling:
> 
> 1. Do any of you have thoughts on the best way to add support for the older V2 contracts? Would it make sense to add a new MintingInterfaceV2 component? And then do something like this in ProjectDetails.tsx:
> 
> ```
> {
>   {
>     "V3": <MintingInterfaceV3 ... />,
>     "V2": <MintingInterfaceV2 ... />
>   }
> }[contractConfig?.VERSION]
> ```
> 

This looks like a solid pattern to me, we don't really have a prescribed way to do this, so this lgtm!

**Comment by matthewjacobson** [comment]
Sorry for the Vercel stuff - I think that was my fault.

So I think I have the last two things figured out. It took a while for me to figure out the erc20 stuff for the V2 contracts but I think I got it sorted out. Please review and let me know if you have any questions!

**Comment by jakerockland** [comment]
Staged deployment is looking good to me!! 🙌 🙌 

https://artblocks-engine-react-96lo14cc4-artblocks.vercel.app/projects

QQ: have you validated that the purchase flow is working as you'd expect with this changeset @matthewjacobson ?

**Comment by matthewjacobson** [comment]
> Staged deployment is looking good to me!! 🙌 🙌
> 
> https://artblocks-engine-react-96lo14cc4-artblocks.vercel.app/projects
> 
> QQ: have you validated that the purchase flow is working as you'd expect with this changeset @matthewjacobson ?

Thanks for taking a look! I am currently in the process of validating the purchase flow - should have a few more changes pushed in a day or two. I am currently only able to test the MinterSetPriceV4 and MinterMerkleV5 minters as that is what has been deployed for the [plottables V3 contracts](https://github.com/ArtBlocks/artblocks-contracts/blob/32738da594e7b9d18e25011b1c7fefa4abb1bda9/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md). Happy to test others out if you want to deploy more on the plottables minter filter :)

**Comment by jakerockland** [comment]
> > Staged deployment is looking good to me!! 🙌 🙌
> > https://artblocks-engine-react-96lo14cc4-artblocks.vercel.app/projects
> > QQ: have you validated that the purchase flow is working as you'd expect with this changeset @matthewjacobson ?
> 
> Thanks for taking a look! I am currently in the process of validating the purchase flow - should have a few more changes pushed in a day or two. I am currently only able to test the MinterSetPriceV4 and MinterMerkleV5 minters as that is what has been deployed for the [plottables V3 contracts](https://github.com/ArtBlocks/artblocks-contracts/blob/32738da594e7b9d18e25011b1c7fefa4abb1bda9/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md). Happy to test others out if you want to deploy more on the plottables minter filter :)

Fantastic 🙏 🙏 

I think that is perfect for now – no need to test with others unless you want to add support for others and test it for Plottables anyway, in which case would be happy to get you additional minters as-needed @matthewjacobson 

cc @ABDruid

**Comment by matthewjacobson** [comment]
Hey @jakerockland - I have verified the purchase flows for the old V2 contract and the new MinterSetPriceV4 and MinterMerkleV5 minters. I included the original logic for MinterDAExpV4 as well but haven't tested it yet. Happy to run through that if I can get it deployed. I think the only other minters I would be interested in supporting for Plottables would be MinterHolderV4 and MinterSetPriceERC20V4 - but no pressure if there is no bandwidth for the deployment work!

Only other comment is related to MinterMerkleV5. I wrote it to hit an api endpoint (.env variable) in order to calculate the merkle proof. Currently this api is something that each partner would have to set up and configure. Two questions:
1) does this seem like a reasonable/generalized solution?
2) would this be something that AB could support natively in the future?

Otherwise let me know if you have any questions or if you want me to add a wallet to the allowlist for the MinterMerkleV5 test project or send some ERC20 tokens for the V2 ERC20 test project.

**Comment by jakerockland** [comment]
> Hey @jakerockland - I have verified the purchase flows for the old V2 contract and the new MinterSetPriceV4 and MinterMerkleV5 minters. I included the original logic for MinterDAExpV4 as well but haven't tested it yet. Happy to run through that if I can get it deployed. I think the only other minters I would be interested in supporting for Plottables would be MinterHolderV4 and MinterSetPriceERC20V4 - but no pressure if there is no bandwidth for the deployment work!

Amazing ser – super appreciate you validating that logic, and no worries on the DA testing. In that case, would you mind updating the README to reflect that the supported minters to reflect the ones that you've tested? I think its OK to remove the EXPDA from that list for now until we get a chance to test on our end. 

Regarding getting you a MinterHolderV4 and MinterSetPriceERC20V4, I'm more than happy to do so! Especially given that you're helping improve the template here which is **super super** appreciated ser. 

Was just doing a bunch of deployments today anyway, so added yours to the list, you can find the full details here (note that you will need to allowlist them on your minterfilter yourself, as I couldn't automate this part given that we've already transferred ownership over to you): https://github.com/ArtBlocks/artblocks-contracts/blob/09efc513c829162c4e7a3444d90b3ef4995eb7b1/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md#minter-deployment

> Only other comment is related to MinterMerkleV5. I wrote it to hit an api endpoint (.env variable) in order to calculate the merkle proof. Currently this api is something that each partner would have to set up and configure. Two questions:
> 
> 1. does this seem like a reasonable/generalized solution?
> 2. would this be something that AB could support natively in the future?

I think this is a reasonable solution for now – would just ask that you add details to the README clarifying that.

That said, regarding the handling of this in the future (why I say "for now"), we do have plans in the near future to move to a world where there is a unified minter suite that can be used across all V3+ core contracts, meaning that there is a shared minterfilter and shared set of minters (concrete implementation wise, not just same contract code wise) which will make it such that the Art Blocks admin/artist experience can be used for things like minter configuration (including merkle proof minter management).

Once this is complete, I _think_ it should be relatively simple on our end to also provide the merkle proof calculation, but will need to talk more with @Jbern16 @yoshiwarab and @ryley-o on that one internally to get a sense of how we might expose that to partners.

tl;dr, we don't have specific immediate plans committed to how we'll approach that once our minter suite re-architecture work is completed, but its not off the table.

**Comment by matthewjacobson** [comment]
Amazing! Thanks so much @jakerockland!

I'll get to work on adding support for those two new minter types and start working on updating the README

**Comment by jakerockland** [comment]
> Amazing! Thanks so much @jakerockland!
> 
> I'll get to work on adding support for those two new minter types and start working on updating the README

Thank **YOU** so much ser! Sounds great to me 🙏 🙏 🙏

**Comment by matthewjacobson** [comment]
Hey @jakerockland - apologies for being out of touch recently, I was having too much fun in Tokyo :) but I'm back now and ready to get back to work!

I saw your message over on #14 - I'll sync those changes in.

With respect to the two other minters you deployed:
1) Is there any documentation on how the MinterHolderV4 is supposed to work?
2) Any chance we can verify and publish the contract source code for MinterSetPriceERC20V4? Currently not able to interact with it on etherscan - https://goerli.etherscan.io/address/0xE79B6D89cc2126d8f1bEb128C748b25cc6D21420#code

**Comment by jakerockland** [comment]
> Hey @jakerockland - apologies for being out of touch recently, I was having too much fun in Tokyo :) but I'm back now and ready to get back to work!

No worry at all – thank you so much @matthewjacobson 💜 🙏 

> 1. Is there any documentation on how the MinterHolderV4 is supposed to work?

Unfortunately we don't have _great_ documentation on this right now, but the tl;dr is that it is intended to be used to gate minting to the holders of another Engine or Flagship project (e.g. gate minting to only Squiggle holders).

More detailed breakdown can be found in the natspec: https://artblocks.github.io/artblocks-contracts/#/contracts/archive/minter-suite/Minters/MinterHolder/MinterHolderV3.sol:MinterHolderV3, though this is pretty verbose.

Happy to answer any questions though that you have to help clarify ser!  

> 2. Any chance we can verify and publish the contract source code for MinterSetPriceERC20V4? Currently not able to interact with it on etherscan - https://goerli.etherscan.io/address/0xE79B6D89cc2126d8f1bEb128C748b25cc6D21420#code

Whoops, my apology I did not realize that the verification initially failed.

Updated:

```
Successfully verified contract MinterSetPriceERC20V4 on Etherscan.
https://goerli.etherscan.io/address/0xE79B6D89cc2126d8f1bEb128C748b25cc6D21420#code
```

**Comment by matthewjacobson** [comment]
Hi @jakerockland!

Okay I have added support for the MinterSetPriceERC20V4 and MinterHolderV4 minters. I can start working on the DA minters once you can get those deployed for me on goerli.

With respect to the upstream merge mentioned in #14 - to be honest I am a bit confused on the best way to go about it. I am definitely not a git wizard and I am getting a bit lost on what exactly I should do. Any guidance would be very much appreciated 💜

**Comment by jakerockland** [comment]
> Okay I have added support for the MinterSetPriceERC20V4 and MinterHolderV4 minters. I can start working on the DA minters once you can get those deployed for me on goerli.

Amazing, will get two exponential DA minters (with and w/o settlement) deployed for you later today and post an update that has updated deployment logs!

> With respect to the upstream merge mentioned in #14 - to be honest I am a bit confused on the best way to go about it. I am definitely not a git wizard and I am getting a bit lost on what exactly I should do. Any guidance would be very much appreciated 💜

If you click on the "Use the command line" hyperlink that I believe you should see under the conflicts message, it should walk you through how to pull down the upstream changes into your branch, and then from there you should just resolve any conflicts by picking the changes that reflect your changeset and work as expected still with your updates.

Screenshot 2023-05-22 at 10 30 03 AM

Let me know how I can help clarify further ser and am happy to help. 🙏

**Comment by matthewjacobson** [comment]
> If you click on the "Use the command line" hyperlink that I believe you should see under the conflicts message, it should walk you through how to pull down the upstream changes into your branch, and then from there you should just resolve any conflicts by picking the changes that reflect your changeset and work as expected still with your updates.
> 
> Screenshot 2023-05-22 at 10 30 03 AM

Looks like I don't have the proper access for this
Screen Shot 2023-05-22 at 19 11 12

In other news - I have updated the readme to explain the project setup and configuration (also added a vercel deployment button), along with some other minor fixes. One major change is I added a new /user route to show the owned tokens for a given wallet. Hope that's okay!

**Comment by jakerockland** [comment]
> > If you click on the "Use the command line" hyperlink that I believe you should see under the conflicts message, it should walk you through how to pull down the upstream changes into your branch, and then from there you should just resolve any conflicts by picking the changes that reflect your changeset and work as expected still with your updates.
> > Screenshot 2023-05-22 at 10 30 03 AM
> 
> Looks like I don't have the proper access for this Screen Shot 2023-05-22 at 19 11 12

Hrm, any chance these instructions work? If not, I'm happy to try to resolve these on our end pre-merge.

Screenshot 2023-05-22 at 5 39 37 PM


> In other news - I have updated the readme to explain the project setup and configuration (also added a vercel deployment button), along with some other minor fixes. One major change is I added a new /user route to show the owned tokens for a given wallet. Hope that's okay!

That is better than OK, its fantastic!

---

@matthewjacobson additional minters deployed for you [here](https://github.com/ArtBlocks/artblocks-engine-react/pull/10):

- https://github.com/ArtBlocks/artblocks-contracts/blob/2a63f35b9ae582ce07ab8ed89942fa4c37e7cc57/packages/contracts/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md#minter-deployment-2
- https://github.com/ArtBlocks/artblocks-contracts/blob/2a63f35b9ae582ce07ab8ed89942fa4c37e7cc57/packages/contracts/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md#minter-deployment-3

**Comment by matthewjacobson** [comment]
@jakerockland looks like my upstream rebase was successful :)

Thanks for those new contracts - I'll get to work on integrating them this week!

**Comment by matthewjacobson** [comment]
@jakerockland integration was quicker than I thought it would be :) I have added support for the MinterDAExpV4 and MinterDAExpSettlementV1 minters. I ran test auctions for each but happy to set up more if you or the team would like to test it out for yourselves. Otherwise I think I am good to merge on my side!

**Review by jakerockland — APPROVED** [review]

**Review by jakerockland — APPROVED** [review]

**Comment by jakerockland** [comment]
Amazing 🙏 🙏  🚢 🚢
